### ISSUE TITLE
os: TestRemoveAllRace failures with "invalid argument" on dragonfly-amd64

### ISSUE THREAD

**Issue by bcmills** [golang/go#50716]
```
--- FAIL: TestRemoveAllRace (0.01s)
    os_test.go:2412: unexpected error: *fs.PathError, "readdirnames /tmp/workdir-host-dragonfly-amd64-master/tmp/issue3518556883/b/b/b/b: readdirent b: invalid argument"
FAIL
FAIL	os	1.031s
```

`greplogs --dashboard -md -l -e 'FAIL: TestRemoveAllRace.*\n(    .*\n)*    .*: invalid argument'`

[2022-01-19T20:07:20-d1640d8/dragonfly-amd64](https://build.golang.org/log/102b5814033e8940752112c700b8fc3eef178686)
[2021-12-02T18:02:46-c5c1955/dragonfly-amd64](https://build.golang.org/log/c8f818efb72ba6c0fa0d6a7ca8a262abe7bbe12a)
[2021-11-23T21:59:51-1ac45e0/dragonfly-amd64](https://build.golang.org/log/484fcb45256a36a187cd8359b1a15da74451209a)
[2019-08-28T02:29:25-731e6fc/dragonfly-amd64](https://build.golang.org/log/f320a41b87b6be7dae90e452c52fcd39f360d1fe)

The test is here:
https://cs.opensource.google/go/go/+/master:src/os/os_test.go;l=2385-2418;drc=c8861432b84157363c84220e6ed71180531bf2b4

The `readdirent` part of that error message appears to come from here:
https://cs.opensource.google/go/go/+/master:src/os/dir_unix.go;l=73;drc=2580d0e08d5e9f979b943758d3c49877fb2324cb

@tuxillo, any idea what might be causing this?

**Comment by tuxillo** [comment]
Is this a new test? Has it happened before? The /tmp mount point in the Gobuilder is tmpfs btw.

**Comment by bcmills** [comment]
This test is not new — see the above links for the failures in the dashboard logs. I see one failure this week, two toward the end of last year, and one from 2019 — so this looks like it used to be a somewhat rare race and is now a more common one (possibly due to scheduling effects from the GC pacing changes in Go 1.18).

`dragonfly` is the only builder exhibiting this failure mode, and the test is not skipped on any other Unix platform.

**Comment by tuxillo** [comment]
Ok, we'll see what we can find out. Thanks for reporting!

**Comment by tuxillo** [comment]
There has been some work that could probably fix this issue, but it's in current master:

https://gitweb.dragonflybsd.org/dragonfly.git/commitdiff/e91c5b2636e1fdd0d9e5170f24e5e39987211a49

whereas the gobuilder is running an older kernel (October, 2021). If I understood correctly, you prefer that we don't upgrade it yet.

**Comment by bcmills** [comment]
Yeah, since we can mitigate this particular failure with a test-skip it's probably better to hold off on upgrading until after the Go 1.18 release. (That way if the system upgrade introduces any other test failures we can diagnose them independent of any further changes in the Go 1.18 release itself.)

**Comment by tuxillo** [comment]
Fine for me. Let's keep this issue opened and let me know when I can do the upgrade.

**Comment by tuxillo** [comment]
Can I upgrade the builder already?

**Comment by bcmills** [comment]
Yes, now would be a great time! 😃

**Comment by tuxillo** [comment]
Done! :)

**Comment by bcmills** [comment]
Looks like that did the trick!

`greplogs --dashboard -md -l -e 'FAIL: TestRemoveAllRace.*\n( .*\n)* .*: invalid argument' --since=2022-03-21`

[no results]
